lib/portage/process.py

```diff
@@ -1370,6 +1375,48 @@ def _start_fork(
     return pid
 
 
+def _start_proc(
+    target: Callable[..., None],
+    args: Optional[tuple[Any, ...]] = (),
+    kwargs: Optional[dict[str, Any]] = {},
+    fd_pipes: Optional[dict[int, int]] = None,
+    close_fds: Optional[bool] = False,
+) -> MultiprocessingProcess:
+    """
+    Execute the target function using multiprocess.Process.
+    If the close_fds parameter is True then NotImplementedError
+    is raised, since it is risky to forcefully close file
+    descriptors that have references (bug 374335), and PEP 446
+    should ensure that any relevant file descriptors are
+    non-inheritable and therefore automatically closed on exec.
+
+    Note that for the "spawn" multiprocessing start method,
+    fd_pipes is sent to the child process asynchronously in a
+    background thread. Therefore, the caller cannot safely
+    close the corresponding file descriptors until after the
+    background thread has finished sending fd_pipes. The
+    current simplest practice is to close fd_pipes file
+    descriptors after waiting for the process to exit.
```

Comment:
I ran the tests with multiprocessing start method spawn, and it looks like some file descriptors are being closed too early, so I need to hunt those down and see what can be done. Maybe an option is needed to have selected file descriptors in fd_pipes closed immediately after they've been sent to the child process.
```
Exception in callback SpawnProcess._main_exit(<Task finishe... descriptor')>)
handle: <Handle SpawnProcess._main_exit(<Task finishe... descriptor')>)>
Traceback (most recent call last):
  File "/usr/lib/python3.12/asyncio/events.py", line 84, in _run
    self._context.run(self._callback, *self._args)
  File "portage/lib/_emerge/SpawnProcess.py", line 198, in _main_exit
    main_task.result()
  File "portage/lib/portage/util/_async/ForkProcess.py", line 112, in _main
    await self.scheduler.run_in_executor(
  File "/usr/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "portage/lib/portage/util/_async/ForkProcess.py", line 103, in _send_fd_pipes
    multiprocessing.reduction.send_handle(
  File "/usr/lib/python3.12/multiprocessing/reduction.py", line 184, in send_handle
    sendfds(s, [handle])
  File "/usr/lib/python3.12/multiprocessing/reduction.py", line 149, in sendfds
    sock.sendmsg([msg], [(socket.SOL_SOCKET, socket.SCM_RIGHTS, fds)])
OSError: [Errno 9] Bad file descriptor
```

Reply:
Maybe the solution is to have ForkProcess created duplicates that it is free to close when it finishes sending them.

Reply:
Yeah, the duplicate fd idea turned out really well, allowing me to revert 69a7c68b45074bec28d6092573b01bb54ac39419.